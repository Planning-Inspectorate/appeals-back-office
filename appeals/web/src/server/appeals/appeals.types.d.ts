import { CurrentPermissionSet } from '#environment/permissions';

export interface CheckboxRadioConditionalHtmlParameter {
	html?: string;
}

export interface CheckboxItemParameter {
	value: string;
	text: string;
	conditional?: CheckboxRadioConditionalHtmlParameter;
	checked?: boolean;
}

export interface SelectItemParameter {
	value: string;
	text: string;
	checked?: boolean;
}

export interface DayMonthYearHourMinute {
	day?: number | string;
	month?: number | string; // 1-based, i.e. January === 1 (Date stores this as 0-based value, eg. Date.getMonth() called on a date in January will return 0)
	year?: number | string;
	hour?: number | string;
	minute?: number | string;
}

export type DocumentVirusCheckStatus = 'not_scanned' | 'scanned' | 'affected';
export type DocumentRowDisplayMode = 'none' | 'number' | 'list';

declare global {
	namespace Express {
		interface Request {
			currentFolder: Schema.Folder;
			currentAppeal: Appeal;
			currentComment: Representation;
			currentRepresentation: Representation;
			apiClient: import('got').Got;
			permissions: CurrentPermissionSet;
		}
	}
}
