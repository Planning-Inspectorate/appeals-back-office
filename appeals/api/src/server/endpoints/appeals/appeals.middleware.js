import { ERROR_MUST_BE_BOOLEAN, ERROR_MUST_BE_VALID_APPEAL_STATE } from '#endpoints/constants.js';
import { APPEAL_CASE_STATUS } from 'pins-data-model';

/** @typedef {import('express').Request} Request */
/** @typedef {import('express').Response} Response */
/** @typedef {import('express').NextFunction} NextFunction */

/**
 * @type {import("express").RequestHandler}
 * @returns {Promise<object|void>}
 */
export const validateHasInspector = async (req, res, next) => {
	const hasInspector = req.query.hasInspector;

	if (hasInspector !== undefined && hasInspector !== 'true' && hasInspector !== 'false') {
		return res.status(400).send({ errors: { hasInspector: ERROR_MUST_BE_BOOLEAN } });
	}

	next();
};

/**
 * @type {import("express").RequestHandler}
 * @returns {Promise<object|void>}
 */
export const validateAppealStatus = async (req, res, next) => {
	const status = req.query.status;
	const validStatuses = [
		APPEAL_CASE_STATUS.ASSIGN_CASE_OFFICER,
		APPEAL_CASE_STATUS.VALIDATION,
		APPEAL_CASE_STATUS.READY_TO_START,
		APPEAL_CASE_STATUS.LPA_QUESTIONNAIRE,
		APPEAL_CASE_STATUS.ISSUE_DETERMINATION,
		APPEAL_CASE_STATUS.COMPLETE,
		APPEAL_CASE_STATUS.INVALID,
		APPEAL_CASE_STATUS.WITHDRAWN,
		APPEAL_CASE_STATUS.CLOSED,
		APPEAL_CASE_STATUS.AWAITING_TRANSFER,
		APPEAL_CASE_STATUS.TRANSFERRED,
		APPEAL_CASE_STATUS.EVENT,
		APPEAL_CASE_STATUS.AWAITING_EVENT,
		//S78
		APPEAL_CASE_STATUS.FINAL_COMMENTS,
		APPEAL_CASE_STATUS.STATEMENTS
	];

	const isValidStatus = typeof status === 'string' ? validStatuses.includes(status) : !status;

	if (!isValidStatus) {
		return res.status(400).send({ errors: { status: ERROR_MUST_BE_VALID_APPEAL_STATE } });
	}
	next();
};
