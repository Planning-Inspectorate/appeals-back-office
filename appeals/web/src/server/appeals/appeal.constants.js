/**
 * @typedef {'appealDetails'|'appellantCase'|'lpaQuestionnaire'|'manageDocuments'|'manageFolder'|'manageRelatedAppeals'|'manageNeighbouringSites'|'ipComments'|'viewIpComment'} ServicePageName
 */

export const paginationDefaultSettings = {
	pageSize: 30,
	firstPageNumber: 1
};

export const textInputCharacterLimits = {
	defaultAddressInputLength: 250,
	defaultInputLength: 300,
	defaultTextareaLength: 1000
};

export const kilobyte = 1000;
export const megabyte = 1000000;
export const gigabyte = 1000000000;
