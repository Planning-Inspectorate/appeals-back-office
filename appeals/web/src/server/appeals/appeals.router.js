import { Router as createRouter } from 'express';
import nationalListRouter from '../appeals/national-list/national-list.router.js';
import personalListRouter from '../appeals/personal-list/personal-list.router.js';
import appealDetailsRouter from './appeal-details/appeal-details.router.js';
import { assertUserHasPermission } from '#app/auth/auth.guards.js';
import { permissionNames } from '#environment/permissions.js';

const router = createRouter();

router.use(
	'/appeals-list',
	assertUserHasPermission(permissionNames.viewCaseList),
	nationalListRouter
);

router.use('/personal-list', personalListRouter);

router.use('/appeal-details', appealDetailsRouter);

export default router;
