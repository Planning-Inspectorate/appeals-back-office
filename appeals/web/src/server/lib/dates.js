import { formatInTimeZone, utcToZonedTime, zonedTimeToUtc } from 'date-fns-tz';
import enGB from 'date-fns/locale/en-GB/index.js';
import { isValid, isBefore, isAfter } from 'date-fns';
import { padNumberWithZero } from '#lib/string-utilities.js';
import { DEFAULT_TIMEZONE } from '@pins/appeals/constants/dates.js';

/**
 * @typedef {import('../appeals/appeals.types.js').DayMonthYearHourMinute} DayMonthYearHourMinute
 */

/**
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {boolean}
 */
export const dateIsValid = (dayMonthYearHourMinute) => {
	const date = new Date(dayMonthYearHourMinuteToISOString(dayMonthYearHourMinute));
	return isValid(date);
};

/**
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {boolean}
 */
export const dateIsInTheFuture = (dayMonthYearHourMinute) => {
	const date = new Date(dayMonthYearHourMinuteToISOString(dayMonthYearHourMinute));

	return isAfter(date, new Date(getTodaysISOString()));
};

/**
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {boolean}
 */
export const dateIsInThePast = (dayMonthYearHourMinute) => {
	const dateISOString = dayMonthYearHourMinuteToISOString(dayMonthYearHourMinute);
	const todaysISOString = getTodaysISOString();

	const date = new Date(dateISOString);
	return isBefore(date, new Date(todaysISOString));
};

/**
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {boolean}
 */
export const dateIsTodayOrInThePast = (dayMonthYearHourMinute) => {
	const dateISOString = dayMonthYearHourMinuteToISOString(dayMonthYearHourMinute);
	const todaysISOString = getTodaysISOString();

	if (dateISOString === todaysISOString) {
		return true;
	}

	const date = new Date(dateISOString);
	return isBefore(date, new Date(todaysISOString));
};

/**
 * @param {string | null | undefined} dateISOString
 * @returns {string}
 */
export function dateISOStringToDisplayTime24hr(dateISOString) {
	if (typeof dateISOString === 'undefined' || dateISOString === null) {
		return '';
	}

	let displayTimeString;

	try {
		displayTimeString = formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'H:mm');
	} catch (e) {
		displayTimeString = '';
	}

	return displayTimeString;
}

/**
 * @param {string | null | undefined} dateISOString
 * @returns {string}
 */
export function dateISOStringToDisplayTime12hr(dateISOString) {
	if (typeof dateISOString === 'undefined' || dateISOString === null) {
		return '';
	}

	let displayTimeString;

	try {
		displayTimeString = formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, `h:mmaaa`);
	} catch (e) {
		displayTimeString = '';
	}

	return displayTimeString;
}

/**
 * returns today's ISO string with today's day, month, year at Europe/london midnight time (00:00)
 * @returns {string}
 */
export function getTodaysISOString() {
	const nowISOString = new Date().toISOString();

	const { year, month, day } = dateISOStringToDayMonthYearHourMinute(nowISOString);

	return dayMonthYearHourMinuteToISOString({
		day,
		month,
		year
	});
}

/**
 * @param {string | null | undefined} dateISOString
 * @param {string} fallback
 * @returns {string}
 */
export function dateISOStringToDisplayDate(dateISOString, fallback = '') {
	if (typeof dateISOString === 'undefined' || dateISOString === null) {
		return fallback;
	}

	let displayDateString;

	try {
		displayDateString = formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'd MMMM yyyy', {
			locale: enGB
		});
	} catch (e) {
		displayDateString = '';
	}

	return displayDateString;
}

/**
 * @param {string | null | undefined} dateISOString
 * @returns {DayMonthYearHourMinute}
 */
export function dateISOStringToDayMonthYearHourMinute(dateISOString) {
	if (typeof dateISOString === 'undefined' || dateISOString === null) {
		return {};
	}

	let day, month, year, hour, minute;

	try {
		day = parseInt(formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'dd'), 10);
		month = parseInt(formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'MM'), 10);
		year = parseInt(formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'yyyy'), 10);
		hour = parseInt(formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'HH'), 10);
		minute = parseInt(formatInTimeZone(dateISOString, DEFAULT_TIMEZONE, 'mm'), 10);
	} catch (e) {
		day = undefined;
		month = undefined;
		year = undefined;
		hour = undefined;
		minute = undefined;
	}

	return { day, month, year, hour, minute };
}

/**
 * Parse the date and time parameters in Europe/London
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {string}
 */
export const dayMonthYearHourMinuteToISOString = (dayMonthYearHourMinute) => {
	if (typeof dayMonthYearHourMinute === 'undefined' || dayMonthYearHourMinute === null) {
		return '';
	}

	const { year, month, day, hour = 0, minute = 0 } = dayMonthYearHourMinute;

	if (
		typeof year === 'undefined' ||
		year === null ||
		typeof month === 'undefined' ||
		month === null ||
		typeof day === 'undefined' ||
		day === null
	) {
		return '';
	}

	const dateStr = `${year}-${padNumberWithZero(month)}-${padNumberWithZero(day)}`;
	const timeStr = `${padNumberWithZero(hour)}:${padNumberWithZero(minute)}`;

	let dateISOString;

	try {
		dateISOString = zonedTimeToUtc(`${dateStr} ${timeStr}`, DEFAULT_TIMEZONE).toISOString();
	} catch (e) {
		dateISOString = '';
	}

	return dateISOString;
};

/**
 * @param {DayMonthYearHourMinute} dayMonthYearHourMinute
 * @returns {string}
 */
export function dayMonthYearHourMinuteToDisplayDate(dayMonthYearHourMinute) {
	if (typeof dayMonthYearHourMinute === 'undefined' || dayMonthYearHourMinute === null) {
		return '';
	}

	return dateISOStringToDisplayDate(dayMonthYearHourMinuteToISOString(dayMonthYearHourMinute));
}

/**
 * @param {string | null | undefined} isoDate
 * @returns {string} Day of the week
 */
export const getDayFromISODate = (isoDate) => {
	if (typeof isoDate === 'undefined' || isoDate === null) {
		return '';
	}
	const dateInZone = utcToZonedTime(isoDate, DEFAULT_TIMEZONE);
	return new Intl.DateTimeFormat('en-GB', { weekday: 'long' }).format(dateInZone);
};
