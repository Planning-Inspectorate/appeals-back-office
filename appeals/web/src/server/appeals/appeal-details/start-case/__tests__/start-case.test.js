import featureFlags from '#common/feature-flags.js';
import { dateISOStringToDisplayDate } from '#lib/dates.js';
import { appealData } from '#testing/appeals/appeals.js';
import { createTestEnvironment } from '#testing/index.js';
import { jest } from '@jest/globals';
import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';

const { app, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';
const appealDataWithoutStartDate = {
	...appealData,
	startedAt: null
};

describe('start-case', () => {
	afterEach(teardown);

	describe('GET /start-case/add', () => {
		it('should render the start case page with the expected content if the appeal type is Householder', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Householder'
				});

			const response = await request.get(`${baseUrl}/1/start-case/add`);

			expect(response.statusCode).toBe(200);

			const element = parseHtml(response.text);

			expect(element.innerHTML).toContain('Start case</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				`Warning</span> Confirming will activate the timetable on ${dateISOStringToDisplayDate(
					new Date().toISOString()
				)}. Start letters will be sent to the relevant parties.`
			);
			expect(unprettifiedElement.innerHTML).toContain('Confirm</button>');
		});

		it('should render the start case page with the expected content if the appeal type is S78 and the S78 hearing feature flag is disabled', async () => {
			featureFlags.isFeatureActive = () => false;

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(`${baseUrl}/1/start-case/add`);

			expect(response.statusCode).toBe(200);

			const element = parseHtml(response.text);

			expect(element.innerHTML).toContain('Start case</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				`Warning</span> Confirming will activate the timetable on ${dateISOStringToDisplayDate(
					new Date().toISOString()
				)}. Start letters will be sent to the relevant parties.`
			);
			expect(unprettifiedElement.innerHTML).toContain('Confirm</button>');
		});

		it('should redirect to the select procedure page if the appeal type is S78 and the S78 hearing feature flag is enabled', async () => {
			featureFlags.isFeatureActive = () => true;

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(`${baseUrl}/1/start-case/add`);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure'
			);
		});

		it('should redirect to the select procedure page if the appeal type is S78 and the S78 hearing feature flag is enabled, passing the backLink query parameter if provided', async () => {
			featureFlags.isFeatureActive = () => true;

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(`${baseUrl}/1/start-case/add?backUrl=/test/back/url`);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure?backUrl=/test/back/url'
			);
		});

		it('should redirect to the select procedure page if the appeal type is S78 and the S78 hearing feature flag is enabled, and the select procedure type page was previously submitted with an option selected', async () => {
			featureFlags.isFeatureActive = () => true;

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const postSelectProcedureResponse = await request
				.post('/appeals-service/appeal-details/1/start-case/select-procedure')
				.send({
					appealProcedure: 'written'
				});

			expect(postSelectProcedureResponse.statusCode).toBe(302);
			expect(postSelectProcedureResponse.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(`${baseUrl}/1/start-case/add?backUrl=/test/back/url`);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure?backUrl=/test/back/url'
			);
		});
	});

	describe('POST /start-case/add', () => {
		it('should redirect to appeal details page', async () => {
			nock('http://test/').get('/appeals/1').reply(200, appealDataWithoutStartDate);
			nock('http://test/').post(`/appeals/1/appeal-timetables`).reply(200, {});

			const response = await request.post(`${baseUrl}/1/start-case/add`).send({});

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe('Found. Redirecting to /appeals-service/appeal-details/1');
		});
	});

	describe('GET /start-case/change', () => {
		it.each([
			['working day', '2024-06-03T22:59:00.000Z', '3 June 2024'],
			['weekend', '2024-06-01T22:59:00.000Z', '3 June 2024']
		])(
			'should render the change start date page with the expected content on a %s',
			async (_, dateString, expectedDisplayDate) => {
				jest
					.useFakeTimers({ doNotFake: ['nextTick', 'setImmediate'] })
					.setSystemTime(new Date(dateString));
				try {
					nock('http://test/').get('/appeals/1').reply(200, appealData);

					const response = await request.get(`${baseUrl}/1/start-case/change`);
					const element = parseHtml(response.text);

					expect(element.innerHTML).toContain('Change start date</h1>');

					const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

					expect(unprettifiedElement.innerHTML).toContain(
						`Warning</span> Confirming will change the start day to ${expectedDisplayDate} and update the case timetable. New start letters will be sent to relevant parties.</strong>`
					);
					expect(unprettifiedElement.innerHTML).toContain('Confirm</button>');
				} finally {
					jest.useRealTimers();
				}
			}
		);
	});

	describe('POST /start-case/change', () => {
		it('should redirect to the appeal details page', async () => {
			nock('http://test/').get('/appeals/1').reply(200, appealData);
			nock('http://test/').post(`/appeals/1/appeal-timetables`).reply(200, {});

			const response = await request.post(`${baseUrl}/1/start-case/change`).send({});

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe('Found. Redirecting to /appeals-service/appeal-details/1');
		});
	});

	describe('GET /start-case/select-procedure', () => {
		it('should render the select procedure page with the expected content', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(
				'/appeals-service/appeal-details/1/start-case/select-procedure'
			);

			expect(response.statusCode).toBe(200);

			const html = parseHtml(response.text).innerHTML;

			expect(html).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			});

			expect(unprettifiedElement.innerHTML).toContain('href="/" class="govuk-back-link">Back</a>');
			expect(unprettifiedElement.innerHTML).toContain(
				'<span class="govuk-caption-l">Appeal 351062 - start case</span>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Appeal procedure</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="written">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="hearing">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="inquiry">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the select procedure page with the expected back link URL, if the backLink query parameter was passed', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(
				'/appeals-service/appeal-details/1/start-case/select-procedure?backUrl=/test/back/url'
			);

			expect(response.statusCode).toBe(200);

			const unprettifiedElement = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			});

			expect(unprettifiedElement.innerHTML).toContain(
				'href="/test/back/url" class="govuk-back-link">Back</a>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'<span class="govuk-caption-l">Appeal 351062 - start case</span>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Appeal procedure</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="written">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="hearing">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="appealProcedure" type="radio" value="inquiry">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		[
			['false', 'select-procedure/check-and-confirm'],
			['true', 'hearing']
		].forEach(([featureFlag, redirectPath]) => {
			describe(`with featureFlagHearingPostMvp set to ${featureFlag}`, () => {
				it('should render the select procedure page with the expected radio option preselected if an appeal procedure is found in the session', async () => {
					nock('http://test/')
						.get('/appeals/1')
						.reply(200, {
							...appealDataWithoutStartDate,
							appealType: 'Planning appeal'
						});

					const selectProcedurePostResponse = await request
						.post('/appeals-service/appeal-details/1/start-case/select-procedure')
						.set('x-feature-flag-hearing-post-mvp', featureFlag)
						.send({
							appealProcedure: 'hearing'
						});

					expect(selectProcedurePostResponse.statusCode).toBe(302);
					expect(selectProcedurePostResponse.text).toBe(
						`Found. Redirecting to /appeals-service/appeal-details/1/start-case/${redirectPath}`
					);

					nock('http://test/')
						.get('/appeals/1')
						.reply(200, {
							...appealDataWithoutStartDate,
							appealType: 'Planning appeal'
						});

					const response = await request.get(
						'/appeals-service/appeal-details/1/start-case/select-procedure'
					);

					expect(response.statusCode).toBe(200);

					const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

					expect(unprettifiedHtml).toContain('Appeal procedure</h1>');
					expect(unprettifiedHtml).toContain(
						'name="appealProcedure" type="radio" value="written">'
					);
					expect(unprettifiedHtml).toContain(
						'name="appealProcedure" type="radio" value="hearing" checked>'
					);
					expect(unprettifiedHtml).toContain(
						'name="appealProcedure" type="radio" value="inquiry">'
					);
				});
			});
		});

		it('should render the select procedure page with no option preselected if the flow was restarted after submitting the select procedure page with an option selected', async () => {
			featureFlags.isFeatureActive = () => true;

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const postSelectProcedureResponse = await request
				.post('/appeals-service/appeal-details/1/start-case/select-procedure')
				.send({
					appealProcedure: 'written'
				});

			expect(postSelectProcedureResponse.statusCode).toBe(302);
			expect(postSelectProcedureResponse.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const getStartCaseAddResponse = await request.get(
				`${baseUrl}/1/start-case/add?backUrl=/test/back/url`
			);

			expect(getStartCaseAddResponse.statusCode).toBe(302);
			expect(getStartCaseAddResponse.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/start-case/select-procedure?backUrl=/test/back/url'
			);

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(
				'/appeals-service/appeal-details/1/start-case/select-procedure'
			);

			expect(response.statusCode).toBe(200);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('Appeal procedure</h1>');
			expect(unprettifiedHtml).toContain('name="appealProcedure" type="radio" value="written">');
			expect(unprettifiedHtml).toContain('name="appealProcedure" type="radio" value="hearing">');
			expect(unprettifiedHtml).toContain('name="appealProcedure" type="radio" value="inquiry">');
			expect(unprettifiedHtml).not.toContain('checked');
		});
	});

	describe('POST /start-case/select-procedure', () => {
		it('should re-render the select procedure page with the expected error message if no radio option was selected', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request
				.post('/appeals-service/appeal-details/1/start-case/select-procedure')
				.send({});

			expect(response.statusCode).toBe(200);

			const html = parseHtml(response.text).innerHTML;

			expect(html).toMatchSnapshot();

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('Appeal procedure</h1>');

			const unprettifiedErrorSummaryHtml = parseHtml(response.text, {
				skipPrettyPrint: true,
				rootElement: '.govuk-error-summary'
			}).innerHTML;

			expect(unprettifiedErrorSummaryHtml).toContain('There is a problem</h2>');
			expect(unprettifiedErrorSummaryHtml).toContain('Select the appeal procedure</a>');
		});

		[
			['false', 'select-procedure/check-and-confirm'],
			['true', 'hearing']
		].forEach(([featureFlag, redirectPath]) => {
			describe(`with featureFlagHearingPostMvp set to ${featureFlag}`, () => {
				it(`should redirect to ${redirectPath} if a radio option was selected`, async () => {
					nock('http://test/')
						.get('/appeals/1')
						.reply(200, {
							...appealDataWithoutStartDate,
							appealType: 'Planning appeal'
						});

					const response = await request
						.post('/appeals-service/appeal-details/1/start-case/select-procedure')
						.set('x-feature-flag-hearing-post-mvp', featureFlag)
						.send({
							appealProcedure: 'hearing'
						});

					expect(response.statusCode).toBe(302);
					expect(response.text).toBe(
						`Found. Redirecting to /appeals-service/appeal-details/1/start-case/${redirectPath}`
					);
				});
			});
		});
	});

	describe('GET /start-case/select-procedure/check-and-confirm', () => {
		it('should render a 500 error page if an appeal procedure is not found in the session', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(
				'/appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			expect(response.statusCode).toBe(500);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('Sorry, there is a problem with the service</h1>');
		});

		it('should render the check details and start case page with the expected content if an appeal procedure is found in the session', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const selectProcedurePostResponse = await request
				.post('/appeals-service/appeal-details/1/start-case/select-procedure')
				.send({
					appealProcedure: 'hearing'
				});

			expect(selectProcedurePostResponse.statusCode).toBe(302);

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.get(
				'/appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			expect(response.statusCode).toBe(200);

			const html = parseHtml(response.text).innerHTML;

			expect(html).toMatchSnapshot();

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('Check details and start case</h1>');
			expect(unprettifiedHtml).toContain('Appeal procedure</dt>');
			expect(unprettifiedHtml).toContain('Hearing</dd>');
			expect(unprettifiedHtml).toContain(
				'href="/appeals-service/appeal-details/1/start-case/select-procedure" data-cy="change-appeal-procedure">Change<span class="govuk-visually-hidden"> Appeal procedure</span></a>'
			);
			expect(unprettifiedHtml).toContain(
				'We’ll start the timetable now and send emails to the relevant parties.</p>'
			);
			expect(unprettifiedHtml).toContain('Start case</button>');
		});
	});

	describe('POST /start-case/select-procedure/check-and-confirm', () => {
		it('should render a 500 error page if an appeal procedure is not found in the session', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const response = await request.post(
				'/appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			expect(response.statusCode).toBe(500);

			const unprettifiedHtml = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;

			expect(unprettifiedHtml).toContain('Sorry, there is a problem with the service</h1>');
		});

		it('should send a post request to the appeal timetables endpoint and redirect to the case details page if an appeal procedure is found in the session', async () => {
			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const selectProcedurePostResponse = await request
				.post('/appeals-service/appeal-details/1/start-case/select-procedure')
				.send({
					appealProcedure: 'hearing'
				});

			expect(selectProcedurePostResponse.statusCode).toBe(302);

			nock('http://test/')
				.get('/appeals/1')
				.reply(200, {
					...appealDataWithoutStartDate,
					appealType: 'Planning appeal'
				});

			const mockAppealTimetablesEndpoint = nock('http://test/')
				.post('/appeals/1/appeal-timetables')
				.reply(200);

			const response = await request.post(
				'/appeals-service/appeal-details/1/start-case/select-procedure/check-and-confirm'
			);

			expect(response.statusCode).toBe(302);
			expect(mockAppealTimetablesEndpoint.isDone()).toBe(true);
		});
	});
});
