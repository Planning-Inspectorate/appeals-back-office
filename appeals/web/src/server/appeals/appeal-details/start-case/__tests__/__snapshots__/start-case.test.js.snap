// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start-case GET /start-case/add/confirmation should render "Case started" confirmation page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-width-container">
        <div class="govuk-grid-row">
            <div class="govuk-grid-column-two-thirds govuk-!-padding-top-2">
                <div class="govuk-panel govuk-panel--confirmation">
                    <h1 class="govuk-panel__title"> Case started</h1>
                    <div class="govuk-panel__body">Case reference number
                        <br><strong>351062</strong>
                    </div>
                </div>
                <p class="govuk-body">Case timetable activated. The relevant parties have been informed.</p>
                <p                 class="govuk-body"><a class="govuk-link" href="/appeals-service/appeal-details/1">Go back to case details</a>
                    </p>
            </div>
        </div>
    </div>
</main>"
`;

exports[`start-case GET /start-case/change/confirmation should render "Start date changed" confirmation page 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-width-container">
        <div class="govuk-grid-row">
            <div class="govuk-grid-column-two-thirds govuk-!-padding-top-2">
                <div class="govuk-panel govuk-panel--confirmation">
                    <h1 class="govuk-panel__title"> Start date changed</h1>
                    <div class="govuk-panel__body">Case reference number
                        <br><strong>351062</strong>
                    </div>
                </div>
                <p class="govuk-body">Case timetable updated. The relevant parties have been informed.</p>
                <p                 class="govuk-body"><a class="govuk-link" href="/appeals-service/appeal-details/1">Go back to case details</a>
                    </p>
            </div>
        </div>
    </div>
</main>"
`;
